File: scripts/verify_config.py
#!/usr/bin/env python3
"""Verify user and project scope config files for safety-net."""

import sys
from pathlib import Path

try:
    from scripts.safety_net_impl.config import ValidationResult, validate_config_file
except ImportError:  # When executed as a script from the scripts/ directory.
    from safety_net_impl.config import (  # type: ignore[no-redef]
        ValidationResult,
        validate_config_file,
    )

_USER_CONFIG = Path.home() / ".cc-safety-net" / "config.json"
_PROJECT_CONFIG = Path(".safety-net.json")

_HEADER = "Safety Net Config"
_SEPARATOR = "═" * len(_HEADER)


def _print_header() -> None:
    print(_HEADER)
    print(_SEPARATOR)


def _print_valid_config(scope: str, path: Path, result: ValidationResult) -> None:
    print(f"\n✓ {scope} config: {path}")
    if result.rule_names:
        print("  Rules:")
        for i, name in enumerate(result.rule_names, 1):
            print(f"    {i}. {name}")
    else:
        print("  Rules: (none)")


def _print_invalid_config(scope: str, path: Path, errors: list[str]) -> None:
    print(f"\n✗ {scope} config: {path}", file=sys.stderr)
    print("  Errors:", file=sys.stderr)
    error_num = 1
    for error in errors:
        for part in error.split("; "):
            print(f"    {error_num}. {part}", file=sys.stderr)
            error_num += 1


def main() -> int:
    """Verify config files and print results."""
    has_errors = False
    configs_checked: list[tuple[str, Path, ValidationResult]] = []

    _print_header()

    if _USER_CONFIG.exists():
        result = validate_config_file(str(_USER_CONFIG))
        configs_checked.append(("User", _USER_CONFIG, result))
        if result.errors:
            has_errors = True

    if _PROJECT_CONFIG.exists():
        result = validate_config_file(str(_PROJECT_CONFIG))
        configs_checked.append(("Project", _PROJECT_CONFIG.resolve(), result))
        if result.errors:
            has_errors = True

    if not configs_checked:
        print("\nNo config files found. Using built-in rules only.")
        return 0

    for scope, path, result in configs_checked:
        if result.errors:
            _print_invalid_config(scope, path, result.errors)
        else:
            _print_valid_config(scope, path, result)

    if has_errors:
        print("\nConfig validation failed.", file=sys.stderr)
        return 1

    print("\nAll configs valid.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/core/verify-config.ts
/**
 * Verify user and project scope config files for safety-net.
 */

import { existsSync, readFileSync, writeFileSync } from "node:fs";
import { resolve } from "node:path";
import {
	getProjectConfigPath,
	getUserConfigPath,
	type ValidationResult,
	validateConfigFile,
} from "./config.ts";

export interface VerifyConfigOptions {
	userConfigPath?: string;
	projectConfigPath?: string;
}

const HEADER = "Safety Net Config";
const SEPARATOR = "═".repeat(HEADER.length);
const SCHEMA_URL =
	"https://raw.githubusercontent.com/kenryu42/claude-code-safety-net/main/assets/cc-safety-net.schema.json";

function printHeader(): void {
	console.log(HEADER);
	console.log(SEPARATOR);
}

function printValidConfig(
	scope: string,
	path: string,
	result: ValidationResult,
): void {
	console.log(`\n✓ ${scope} config: ${path}`);
	if (result.ruleNames.size > 0) {
		console.log("  Rules:");
		let i = 1;
		for (const name of result.ruleNames) {
			console.log(`    ${i}. ${name}`);
			i++;
		}
	} else {
		console.log("  Rules: (none)");
	}
}

function printInvalidConfig(
	scope: string,
	path: string,
	errors: string[],
): void {
	console.error(`\n✗ ${scope} config: ${path}`);
	console.error("  Errors:");
	let errorNum = 1;
	for (const error of errors) {
		for (const part of error.split("; ")) {
			console.error(`    ${errorNum}. ${part}`);
			errorNum++;
		}
	}
}

function addSchemaIfMissing(path: string): boolean {
	try {
		const content = readFileSync(path, "utf-8");
		const parsed = JSON.parse(content) as Record<string, unknown>;

		if (parsed.$schema) {
			return false;
		}

		const updated = { $schema: SCHEMA_URL, ...parsed };
		writeFileSync(path, JSON.stringify(updated, null, 2), "utf-8");
		return true;
	} catch {
		return false;
	}
}

/**
 * Verify config files and print results.
 * @returns Exit code (0 = success, 1 = errors found)
 */
export function verifyConfig(options: VerifyConfigOptions = {}): number {
	const userConfig = options.userConfigPath ?? getUserConfigPath();
	const projectConfig = options.projectConfigPath ?? getProjectConfigPath();

	let hasErrors = false;
	const configsChecked: Array<{
		scope: string;
		path: string;
		result: ValidationResult;
	}> = [];

	printHeader();

	if (existsSync(userConfig)) {
		const result = validateConfigFile(userConfig);
		configsChecked.push({ scope: "User", path: userConfig, result });
		if (result.errors.length > 0) {
			hasErrors = true;
		}
	}

	if (existsSync(projectConfig)) {
		const result = validateConfigFile(projectConfig);
		configsChecked.push({
			scope: "Project",
			path: resolve(projectConfig),
			result,
		});
		if (result.errors.length > 0) {
			hasErrors = true;
		}
	}

	if (configsChecked.length === 0) {
		console.log("\nNo config files found. Using built-in rules only.");
		return 0;
	}

	for (const { scope, path, result } of configsChecked) {
		if (result.errors.length > 0) {
			printInvalidConfig(scope, path, result.errors);
		} else {
			if (addSchemaIfMissing(path)) {
				console.log(`\nAdded $schema to ${scope.toLowerCase()} config.`);
			}
			printValidConfig(scope, path, result);
		}
	}

	if (hasErrors) {
		console.error("\nConfig validation failed.");
		return 1;
	}

	console.log("\nAll configs valid.");
	return 0;
}
